core/include/scipp/core/sizes.h

```diff
@@ -8,49 +8,80 @@
 
 #include "scipp-core_export.h"
 #include "scipp/common/index.h"
-#include "scipp/core/dimensions.h"
+#include "scipp/common/span.h"
 #include "scipp/core/slice.h"
 #include "scipp/units/dim.h"
 
 namespace scipp::core {
 
-/// Sibling of class Dimensions, but unordered.
-class SCIPP_CORE_EXPORT Sizes {
-public:
-  Sizes() = default;
-  Sizes(const Dimensions &dims);
-  Sizes(const std::unordered_map<Dim, scipp::index> &sizes) : m_sizes(sizes) {}
+constexpr int32_t NDIM_MAX = 6;
 
-  bool operator==(const Sizes &other) const;
-  bool operator!=(const Sizes &other) const;
+class Dimensions;
 
-  bool contains(const Dim dim) const noexcept {
-    return m_sizes.count(dim) != 0;
+/// Small (fixed maximum size) and stable (preserving key order) map
+template <class Key, class Value, int16_t Capacity>
```

Comment:
This comes from the original goal to fit `class Dimensions` into 64 Byte, at a time when we were having many more instances of the class, for defining nested structures required for event data. Since `Dim` is 16 bit we had `2 + 6 * 2 + 6 * 8`. So 32 bit would have been ok as well, not sure why I chose 16 bit.

Now this should be considerably less relevant and I think it is likely that we will increase NDIM_MAX in the foreseeable future, in particular in light of new operations such as `convolve` which essentially double the number of (temporarily used) dimensions.